### DISCUSSION TITLE
Adding ONE favorite  via API overwrites all "My Places"

### DISCUSSION THREAD

**Discussion by ThisIsEd** [osmandapp/OsmAnd#19633] [Q&A]
I am a beginner on Android Studio. Tried  to add a favorite via API with unexpected results:

        loadButton.setOnClickListener(new View.OnClickListener() {

            public void onClick(View view) {

                // Call the method to add a favorite location to OsmAnd
                addFavoriteToOsmAnd(37.728180, -122.477020,
                        "Marker via API",
                        "3251 20th Ave #187, San Francisco, CA 94132",
                        "Restaurants",
                        "blue",
                        true);

The snippet above did add the favorite entry as expected but did so at the cost of overwriting (erasing) any and all previously existing favorite entries.

I was unable to find a way to use a specific icon shape for a favorite marker in the API documentation.

What I want to do is to:
Merge:
By adding a non existing favorite marker without erasing the existing ones.
Delete:
By deleting an existing favorite marker without erasing the existing ones.
Update:
By editing an existing favorite marker's description without erasing the existing ones.

**Will paypal anyone with the knowledge that will assist me to achieve the above on an app. Please let me know how much and the ETA.** 

App needs to be backwards compatible with Android 9

**Comment by vshcherb** [comment]
As far as I can say, OsmAnd API demo doesn't erase existing Favorites on adding new one (https://download.osmand.net/latest-night-build/OsmAnd-api-sample.apk). So if the sample works as expected, you might need to consult the source code of the samples - https://github.com/osmandapp/osmand-api-demo/tree/master/OsmAnd-api-sample

**Reply by ThisIsEd** [reply]
Thank you for finally replying. I experienced this by using intents: The entire favourites.gpx gets overwritten/written anew, that's what takings place to the best of my recollection.  I haven't tinkered with this since when I posted this but work on it again this weekend. Basically, intents, when creating WPT entries, do not "merge" with existing WPT entries in the existing favourtes.gpx file. Hence the "overwriting issue".   

Aside from the issue described above,  I noticed that nowhere in the SDK documentation I could find any reference to creating a new favorite point via intent or otherwise that would include assigning a specific icon (like a star vs any other icon shape or background). Basically, each WPT created by the actual Osmand standalone app, was more rich in features/parameters than their SDK created counterparts. I am extremely interested in this feature of being able to control the appearance of resulting icons in the map with as many features/parameters as in the standalone version of the app. Please advice regarding the SDK. 

PS: 
Incidentally, I drive an average of 60 ~ 100 miles a  day for work with an average of 15 ~ 25+ stops per day and experience very inaccurate routing/directions to certain areas at least twice a week. I'd be very happy to provide valuable data on these items to improve the "routing algorithm" or whatever it's called. I'd just prefer to not do that in a public forum, but if you were by any means interested in troubleshoot routing issues in Osmand  (which I've noticed have been reported all the way from Australia altogether) I could provide tons of fresh and abundant data for these debugging purposes at least once a week. (If you'd like me to do that).

I'd love to learn a bit more about this project to see if you could use any form of contribution to it. As I said in my original post, a financial contribution in exchange for assistance with the app I need to put together for my work is definitely on the table on my end. What I am working towards in my very little free time, includes access (via intent or via API or otherwise) to current Osmand capabilities so there should be nothing "extra" to write other than clear paths to access its current capabilities. Now, I use Android, but it would be however very interesting to learn if there will be a way to do this on iOS devices in the near future.

Once again, I look forward overcoming this initial impasse and hopefully being able to provide whatever contribution to this project you see fit should we be able to get some traction regarding my API questions/concerns.

Best regards,

Ed

**Reply by vshcherb** [reply]
- Yes entire favorites.gpx get's overwritten cause all points stored in same GPX.Modifying 1 point saves whole file
- Specific icon is part of additional tags and it's not supported by API (public class AFavorite implements Parcelable )
- Method needs to change to support it
```
boolean addFavorite(double latitude, double longitude, String name, String category, String description, String colorTag, boolean visible) {
		point.setColor(color);
...
```

As for now we don't have available developers to change API and we mostly work B2B around SDK. So I don't think it's viable approach. 
As I see some people do automation with Tasker app or may be very soon there will be AI agents that could automate some actions.

**Reply by ThisIsEd** [reply]
How many hours do you think it would take to make the API match the capabilities of the standalone app, and how much would you charge per hour to get this done? Also, what would be a reasonable ETA?

**Reply by ThisIsEd** [reply]
I might be ab;e to afford it. If you'd like to address this in private you might be able to look at support request 59227 and you'll see my email there
